### PR TITLE
Add the ability to specify a lock file

### PR DESCRIPTION
**Pull request by lassizci**
I came around the issue in the old repo: https://github.com/roboll/helmfile/issues/779 while trying to achieve exactly the same thing. So I would like to be able to apply updates in one environment first and then rollout the same changes to the rest of the environments.

So would you accept this kind of change? What else would be good to have in terms of testing / documentation?

### CODE REVIEW

**Comment by yxxhero** [comment]
@lassizci Please fix DCO issue.

**Comment by yxxhero** [comment]
@lassizci need more e2e test.

**Comment by lassizci** [comment]
> @lassizci need more e2e test.

Allright! Which aspects would you recommend? Looking at the current ones, there doesn't seem to be lock file related e2e tests, but I could perhaps try to craft something together, if that's what we are looking for.

File: docs/advanced-features.md
Comment (mumoshu) [inline docs/advanced-features.md:331]: Thanks for the PR!
One thing I noticed is that you should never use a helmfile template within the same fragment that defines `environments`. See https://github.com/helmfile/helmfile/blob/main/docs/proposals/towards-1.0.md#the-changes-in-10
This should ideally be written something like the below:

```suggestion
```yaml
environments:
  staging:
  production

---
lockFile: .helmfile.{{ .Environment.Name}}.lock

releases:
- name: myapp
  chart: charts/myapp
```
```

**Comment by yxxhero** [comment]
> > @lassizci need more e2e test.
> 
> Allright! Which aspects would you recommend? Looking at the current ones, there doesn't seem to be lock file related e2e tests, but I could perhaps try to craft something together, if that's what we are looking for.

yes. that's true.

**Comment by yxxhero** [comment]
@lassizci please fix DCO issue.

**Comment by yxxhero** [comment]
@lassizci please squash your commits.

**Comment by lassizci** [comment]
Now, with a e2e test.

**Comment by yxxhero** [comment]
@lassizci I  have a question.
there is a raw chart in chart dir. why add raw-0.0.1? Thanks very much.

**Comment by lassizci** [comment]
> @lassizci I have a question.
> there is a raw chart in chart dir. why add raw-0.0.1? Thanks very much.

The purpose of lock file is to freeze a chart to a specific version. To be able to test the lock file is effective, I thought multiple versions of the same chart would be needed. Would there be some other way to provide multiple versions of the same chart via local chart repo, than duplicating the chart?

**Comment by yxxhero** [comment]
@lassizci ok. I see.
I think moving raw to raw-0.1.0 will be more clear.

@mumoshu WDYT?

**Comment by lassizci** [comment]
> @lassizci ok. I see. I think moving raw to raw-0.1.0 will be more clear.
Done.

File: pkg/app/desired_state_file_loader.go
Comment (yxxhero) [inline pkg/app/desired_state_file_loader.go:40]: I don't see the assignment of this field. Are you sure it is useful? @lassizci

Comment (lassizci) [reply pkg/app/desired_state_file_loader.go:40]: There seems to be existing convention to have the fields in the [desiredStateLoader](https://github.com/helmfile/helmfile/blob/44bf5590d9a574c7bdf95272cdebef2ee9e25a35/pkg/app/desired_state_file_loader.go#L24-L41) regardless of whether they are assigned or not.


this: https://github.com/helmfile/helmfile/blob/44bf5590d9a574c7bdf95272cdebef2ee9e25a35/pkg/app/desired_state_file_loader.go#L168

cloud also be initialized by just using empty string, such as:
```go
c := state.NewCreator(a.logger, a.fs, a.valsRuntime, a.getHelm, a.overrideHelmBinary, a.remote, a.enableLiveOutput, "")
```
but then it wouldn't follow the existing style and I think `""` is less obvious, which argument are we passing.

Comment (yxxhero) [reply pkg/app/desired_state_file_loader.go:40]: @mumoshu WDYT?

**Review by mumoshu — APPROVED** [review]
Sorry for the delay, and thanks for your contribution! LGTM overall, but allow me to change lockFile to lockFilePath before merging, so that it is more clear that it is not about toggling the lock feature.

**Comment by yxxhero** [comment]
@mumoshu do you need me to do something?

**Comment by mumoshu** [comment]
@yxxhero If you could update code to use lockFilePath instead in favor of me, that would be awesome- I'll review the update asap and can approve/merge this earlier!

**Comment by yxxhero** [comment]
@mumoshu ok. I will see this PR later.

**Comment by yxxhero** [comment]
@lassizci @mumoshu lassizci:configurable-lockfile is not helmfile branch. It's not very convenient to change the code.

@lassizci could you update the code?

**Comment by lassizci** [comment]
@yxxhero  @mumoshu Thanks for the feedback, sounds like reasonable thing, so `lockFile` now renamed to `lockFilePath`.

**Comment by mumoshu** [comment]
@yxxhero In case you missed that, maintainers of pull request base branches should be able to push and even force-push heads of pull requests, even though heads are in other repositories. FWIW, I usually use `gh pr checkout $PR_NUM`, make commits, then run `git push` without args to push commits to the pull request head.

**Review by mumoshu — APPROVED** [review]
LGTM. Thank you so much for your patience and contribution @lassizci!

**Comment by yxxhero** [comment]
@mumoshu got it. Thanks very much. great idea.

**Pull request merged**
